Terms

cgalarza opened the pull request.

Commit:
Cleaning up gemfile

Commit:
Removing rspec config file

Commit:
Adding description to rake task

Commit:
Adding configuration to prevent the loading of term-ansicolor, which when loaded to our test environment does not allow us to test our Term model.

Commit:
Adding Term model and database migrations. Creating relationship between Vocabularies and Terms.

Commit:
Adding Rails.application.config.local_uri_host configuration variable.

Commit:
Adding test for Term model

Commit:
Adding Coveralls badge

Commit:
Updating vocabulary factory to relate to term factory

Commit:
Loading classes and modules in /lib

Commit:
Indexing terms into solr core when they are created/updated/deleted.

Commit:
Updates to API specification

Commit:
Adding local_uri_host configuration for development environment

Commit:
Only creating one URIService.solr object. Adding helper for solr suffix

Commit:
Adding routes and controllers for /term endpoint; adding new module to centralize json rendering.

Commit:
Adding json-spec to test json responses

Commit:
Serialize Term object or hash into fields needed for api response

Commit:
Updating solr configuration to create two response handlers, /select and /search.

The /search endpoint should only be used when using the q parameter to search. For filtering the /select endpoint will be faster.

Commit:
Adding tests for term searches

Commit:
Fix to Term json serialization

Commit:
Adding search tests

Commit:
Adding CRUD test for /terms endpoint

Commit:
Remove temporary test

Commit:
Adding test to make sure only one Solr object is created

Commit:
Adding validation for vocabulary string key

elohanlon commented on app/controllers/v1/terms_controller.rb:
```diff
@@ -0,0 +1,123 @@
+module V1
+  class TermsController < ApplicationController
+    # Check that vocabulary is valid, if not should render and error
```
Is this a TODO note?

cgalarza replied:
Yea, this is a TODO note. I can add `TODO:` in front of it if that's helpful.

elohanlon replied:
Thanks! I'm in the habit of periodically searching apps for the string "TODO" to find all TODOs, so that would be great.

elohanlon commented on app/controllers/v1/terms_controller.rb:
```diff
@@ -0,0 +1,123 @@
+  class TermsController < ApplicationController
+    # Check that vocabulary is valid, if not should render and error
+
+    ## force json type
```
Also a TODO note?

cgalarza replied:
Yes, also a TODO note.

elohanlon commented on app/controllers/v1/terms_controller.rb:
```diff
@@ -0,0 +1,123 @@
+
+    # PATCH /vocabularies/:string_key/terms/:uri
+    def update
+      term = Term.find_by(vocabulary: vocabulary, uri: params[:uri]) # should include querying for vocab
```
I'm not sure what this comment means (# should include querying for vocab).

cgalarza replied:
I think that was just to remind me that I need to query by vocab and uri, it can probably be deleted.

elohanlon replied:
Oh okay

elohanlon commented on app/controllers/v1/terms_controller.rb:
```diff
@@ -0,0 +1,123 @@
+        term.assign_attributes(update_params) # updates, but doesn't save.
+
+        custom_fields.each do |f, v|
+          term.add_custom_field(f, params.fetch(f, nil))
```
Would this end up clearing a custom field value if it's not supplied as part of an update? e.g. If this is the `names` vocabulary and someone doesn't supply `name_type` as part of the update, would this line run `term.add_custom_field('name_type', nil)` and clear out an existing `name_type` value?

cgalarza replied:
I think you have a good point there. It would clear out the value. I think I was trying to allow for the clearing of values if a user sends them through as null. But I can check for the presence of the value in the params hash before setting the custom field and that should fix the problem. I'll can tests for this too.

elohanlon replied:
Perfect, thanks!

cgalarza replied:
Added a commit that fixes this and added a test for it.

elohanlon commented on app/models/term.rb:
```diff
@@ -0,0 +1,105 @@
+    end
+  end
+
+  def add_custom_field(field, value)
```
It might be more accurate to call this `set_custom_field` because it replaces an existing value rather than adding.

cgalarza replied:
Sure.

elohanlon commented on app/models/term.rb:
```diff
@@ -0,0 +1,105 @@
+  private
+
+    def add_uuid
+      self.uuid = SecureRandom.uuid unless uuid
```
Should this raise an error instead of silently failing?  If anything is calling add_uuid twice for a term, that would be a problem.

cgalarza replied:
I am a little confused by your comment. I am checking whether or not a uuid is set because I am allowing for a uuid to be set at creation. That way when we migrate the uuid can stay the same for local terms.

cgalarza replied:
We decided that this method will only set the uuid for new records, if the record has already been persisted an error will be raised. UUID validation should also be added.

elohanlon commented on app/models/term.rb:
```diff
@@ -0,0 +1,105 @@
+      when TEMPORARY
+        self.uri = URI(TEMPORARY_URI_BASE + Digest::SHA256.hexdigest(self.vocabulary.string_key + self.pref_label)).to_s
+      when EXTERNAL
+        if self.uri.nil? || !self.uri.match?(VALID_URI_REGEX)
```
It feels sort of weird to have a method called `add_uri` do uri validation for a uri that's already set.  I understand why you're bundling the on-create uri logic into one method, but I think this method might be trying to do too much (and it probably shouldn't run on creation of an external term).

cgalarza replied:
Hmm. Well I still need to validate the uri for external terms so yea, I did group it all together. What would you suggest?

cgalarza replied:
We decided to move out validation of uri from this setter method.

elohanlon commented on app/models/term.rb:
```diff
@@ -0,0 +1,105 @@
+      end
+    end
+
+    def add_uri_hash
```
This could just be me being picky about names, but this feels more like a "set" kind of method rather than an "add" kind of method.  I have a strong association between adding and appending.

cgalarza replied:
We can change it to `set_uri_hash`.

elohanlon commented on app/models/term.rb:
```diff
@@ -0,0 +1,105 @@
+      end
+    end
+
+    def update_solr # If this is unsuccessful the solr core will be out of sync
```
I guess if a solr delete or add call fails, this method would pass along any raised solr exception?  I'd like the client to at least know that something went wrong during the overall save process, even if it's via the after_commit `update_solr` step raising an exception.  Probably fine as is, but I just wanted to check.

cgalarza replied:
Yea, I think if this failed it would raise an error. I am going to add a handler that catches all errors and returns them to the user a generic 500 errors. Do you think we should throw a more specific error for errors raised by this method?

elohanlon replied:
Re in-person conversation: We want to log all errors, but it's fine to do a controller-wide catch of errors and pass a generic error message like "an unexpected error occurred" to the user.

cgalarza replied:
Adding this to list of things TODO.

elohanlon commented on app/models/vocabulary.rb:
```diff
@@ -1,6 +1,15 @@
 class Vocabulary < ApplicationRecord
-  validates :string_key, presence: true, uniqueness: true
+  has_many :term, dependent: :destroy
```
```suggestion
  has_many :terms, dependent: :destroy
```
I think that this should work with the plural form (and that's what I'm used to from the Rails docs).

cgalarza replied:
Okay, will change. That makes a lot of sense

elohanlon commented on app/models/term.rb:
```diff
@@ -0,0 +1,105 @@
+    elsif !vocabulary.custom_fields.keys.include?(field)
+      errors.add(:custom_field, "cannot add #{field} because it is not defined by parent vocabulary")
+    else
+      self.custom_fields[field] = value
```
Since we define `data_type` for custom fields, we should probably do data_type validation here (i.e. cast params, which may come in as all-strings, to their correct types --- and then make sure that the casting didn't change their values like how `'some string'.to_i == 0`.

cgalarza replied:
I think if we want to enforce data type, we should enforce it in the json. We can check the json type of a parameter coming in. I worry about checking if the values changed because `'1' == 1` will not be true.

cgalarza replied:
Re: In person conversation, we decided to do this, but it wont get done as part of this pull request.

elohanlon commented on config/environments/development.rb:
```diff
@@ -51,4 +51,6 @@
   # routes, locales, etc. This feature depends on the listen gem.
   config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+
+  config.local_uri_host = 'localhost:3000'
```
What if I don't run my local webserver on `localhost:3000`?  (I run local Hyacinth on 3000 and would point to localhost:3001 for local URI Service).  My point is that maybe that's actually a good use case for encoding this value in a file that isn't committed to the repo (e.g. uri_service.yml, or somewhere else).  I know we talked about this a while back and I didn't have a good enough justification for a separate file back then, but I feel like I have a concrete use case now.

cgalarza replied:
okay, we can move it to `uri_service.yml`. It also didn't really make sense at the time because it would be the only thing in that config file, but now we have other candidates (api keys, etc).

elohanlon commented on lib/uri_service.rb:
```diff
@@ -0,0 +1,19 @@
+    'boolean' => '_bi'
+  }.freeze
+
+  # TODO: Write test to make sure only one object is created.
```
Okay to delete this TODO?  Looks done?

cgalarza replied:
Oh yea, i did do that lol

elohanlon commented on lib/uri_service/json.rb:
```diff
@@ -0,0 +1,66 @@
+          term_type:  document.fetch(:term_type, nil)
+        }
+
+        if vocabulary = Vocabulary.find_by(string_key: document[:vocabulary])
```
```suggestion
        # TODO: Eventually create a Vocabulary lookup cache so we don't do a db query every time we format a Term/Hash as JSON
        if vocabulary = Vocabulary.find_by(string_key: document[:vocabulary])
```

elohanlon commented on lib/uri_service/solr.rb:
```diff
@@ -0,0 +1,81 @@
+
+    # Returns configuration options set in solr.yml
+    def connection_config
+      Rails.application.config_for(:solr).symbolize_keys
```
I didn't know about `Rails.application.config_for` before now! Handy!

elohanlon committed:
Adding TODO

Co-Authored-By: cgalarza <cmg2228@columbia.edu>

elohanlon commented on lib/uri_service/solr.rb:
```diff
@@ -0,0 +1,81 @@
+    # @param Hash json document to be added to solr
+    def add(doc)
+      connection.add(doc)
+      connection.commit # probably need to make this optional or leverage soft commit
```
I was thinking more about this actually.  We don't want to commit constantly, but we do want to always make sure that our retrieved data is 100% up to date.  This leaves us with two options:

1. Always pull term data from the database.  i.e. Do a solr search (and don't store anything in the solr index other than uuids), then use that list of returned UUIDs to do *one* db query to retrieve all of the terms in that UUID list.   I'm hoping that a specific db query based on a list of UUIDs will be really fast.  If it gets slow, we can consider an intermediate document store later, which could either have a limited size or could potentially store 100% of terms if that's not unreasonable.

2. Use solr for searching and and storing data, but introduce an extra non-solr caching layer for recently updated terms (maybe the last 60 seconds worth of updated terms) and query this cache when returning term search results to ensure that everything is up to date, and fall back to the solr data when something isn't found in the cache (which implies that it wasn't updated recently).  I don't love the additional complexity of this, but it could work.  Maybe backed by redis or something.

What do you think?

cgalarza replied:
Wouldn't searches not reflect the changes made if we don't commit them?

cgalarza replied:
I had an idea about this. Maybe we can make all filter queries database queries since we are matching by exact string. We can use solr to make request that use the `q` parameter, because that does fuzzy searching and what not. That way only queries that use the q parameter might be partially out of sync. We can talk about it later in person.

elohanlon commented on lib/uri_service/solr.rb:
```diff
@@ -0,0 +1,81 @@
+    # Deleting term based on uuid (solr primary key)
+    #
+    # @param String uuid
+    def delete(uuid)
```
```suggestion
    def delete(uuid, commit = true)
```

elohanlon replied:
If we ever want to do a programmatic mass deletion, being able to pass commit = false here would be useful.

cgalarza replied:
I am not going to pull these changes in because I want to make it part of the change that adds a configuration setting about autocommiting.

elohanlon commented on lib/uri_service/solr.rb:
```diff
@@ -0,0 +1,81 @@
+    # @param String uuid
+    def delete(uuid)
+      connection.delete_by_query('uuid:' + solr_escape(uuid))
+      connection.commit # if commit
```
```suggestion
      connection.commit if commit
```

elohanlon commented on lib/uri_service/solr_params.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+
+    def fq(field, value)
+      @parameters[:fq] << "#{field}:\"#{value}\"" unless value.nil? ## should probably escape
```
```suggestion
      @parameters[:fq] << "#{field}:\"#{URIService.solr.solr_escape(value)}\"" unless value.nil?
```
```suggestion
      @parameters[:fq] << "#{field}:\"#{value}\"" unless value.nil? ## should probably escape
```

elohanlon replied:
```suggestion
      @parameters[:fq] << "#{field}:\"#{URIService.solr.solr_escape(value)}\"" unless value.nil?
```

elohanlon commented on lib/uri_service/solr_params.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+
+    def q(query)
+      @parameters[:q] = query unless query.nil?
```
Maybe we want to escape query with `URIService.solr.solr_escape(query)`?  We should think/talk about that.

cgalarza replied:
yea, i think we talked about it earlier, we should be escaping unless we want to allow users more control over querying. I forgot to add that in

elohanlon commented on lib/uri_service/solr_params.rb:
```diff
@@ -0,0 +1,42 @@
+
+    [:vocabulary, :authority, :uri, :pref_label, :alt_label, :term_type].each do |term_attribute|
+      define_method term_attribute do |value|
+        fq(term_attribute.to_s, value)
```
Might also want to use URIService.solr.solr_escape(value) here (in case the literal value has quotes, parens, etc.).

cgalarza replied:
the updated `fq` method should take care of that, right?

elohanlon commented on spec/factories/term.rb:
```diff
@@ -0,0 +1,37 @@
+    factory :temp_term do
+      pref_label { 'Yeti' }
+      alt_label { ['Big Foot'] }
+      term_type { 'temporary' }
```
You'll make it into the LOC NAF one day, Yeti.

elohanlon commented on spec/fixtures/solr_cores/uri_service_solr6/conf/schema.xml:
```diff
@@ -64,22 +68,15 @@
-
-
+  <copyField source="pref_label" dest="pref_label_ssort" />
+  <copyField source="alt_label" dest="alt_label_ssort" />
```
It might not make sense to be able to sort by alt label, since it's multivalued.  I'm actually not sure how that would work in solr.  We may want to remove this copyField.

cgalarza replied:
Oh right, you can't sort by a multivalued field. Will remove.

Commit:
Updating comments

elohanlon commented on spec/models/term_spec.rb:
```diff
@@ -0,0 +1,156 @@
+    let(:term) { FactoryBot.create(:local_term) }
+
+    it 'add local term uri' do
+      expect(term.uri).to start_with 'https://example.com/term/'
```
Do you mind adding an additional check here that the term starts with `Rails.application.config.local_uri_host`? That would be informative for anyone looking at the tests who doesn't know the backing logic as well.

cgalarza replied:
Sure, tho it would be a different look up if we store that in the uri_service.yml.

Commit:
Using pural for has_many relationship

elohanlon commented on spec/models/term_spec.rb:
```diff
@@ -0,0 +1,156 @@
+
+    it 'fails if term_type is nil' do
+      expect {
+        FactoryBot.create(:external_term, term_type: '')
```
This is testing an empty string term_type rather than a nil value.

cgalarza replied:
Updated this test to test for nil and renamed it.

elohanlon commented on spec/requests/v1/term_crud_spec.rb:
```diff
@@ -0,0 +1,279 @@
+      end
+    end
+
+    context 'when invalid uri' do
```
```suggestion
    context 'when attempting to update a uri that is not associated with a known term' do
```
I wanted to change this just because the URI itself isn't an invalid URI.

elohanlon commented on spec/requests/v1/term_crud_spec.rb:
```diff
@@ -0,0 +1,279 @@
+
+    context 'when invalid uri' do
+      before do
+        patch "/api/v1/vocabularies/mythical_creatures/terms/#{CGI.escape('https://example.com/not/valid')}", params: {
```
```suggestion
        patch "/api/v1/vocabularies/mythical_creatures/terms/#{CGI.escape('https://example.com/not/known')}", params: {
```

elohanlon commented on spec/requests/v1/term_crud_spec.rb:
```diff
@@ -0,0 +1,279 @@
+      end
+    end
+
+    context 'when invalid uri' do
```
```suggestion
    context 'when attempting to delete a uri that is not associated with a known term' do
```

elohanlon commented on spec/requests/v1/term_crud_spec.rb:
```diff
@@ -0,0 +1,279 @@
+
+    context 'when invalid uri' do
+      before do
+        delete '/api/v1/vocabularies/mythical_creatures/terms/http%3A%2F%2Fid.worldcat.org%2Ffast%2Fnot_valid%2F'
```
```suggestion
        delete '/api/v1/vocabularies/mythical_creatures/terms/http%3A%2F%2Fid.worldcat.org%2Ffast%2Fnot_known%2F'
```

cgalarza replied:
I am merging these test fixes into one commit.

elohanlon commented on spec/requests/v1/term_filtering_spec.rb:
```diff
@@ -0,0 +1,151 @@
+    FactoryBot.create(:temp_term, vocabulary: unicorn.vocabulary)
+  end
+
+  context 'with no filters' do
```
hash tag #nofilter :P

elohanlon commented on spec/requests/v1/vocabularies_spec.rb:
```diff
@@ -124,15 +131,15 @@
       before do
         FactoryBot.create(:vocabulary)
-        delete '/api/v1/vocabularies/subjects'
+        delete '/api/v1/vocabularies/mythical_creatures'
```
For this test, you may want to check for the presence of the mythical_creatures vocabulary before deleting it, just to prove that it actually existed before the delete succeeded.

elohanlon commented on spec/requests/v1/vocabularies_spec.rb:
```diff
@@ -124,15 +131,15 @@
 
       it 'removes vocabulary from database' do
-        expect(Vocabulary.find_by(string_key: 'subjects')).to be nil
+        expect(Vocabulary.find_by(string_key: 'mythical_creatures')).to be nil
```
For this test, you may want to check for the presence of the mythical_creatures vocabulary before deleting it, just to prove that it actually existed before the delete succeeded.

cgalarza replied:
instead of checking for the existence, i did this:
```ruby
    context 'when deleting vocabulary' do
      let(:vocabulary) { FactoryBot.create(:vocabulary) }

      before do
        delete "/api/v1/vocabularies/#{vocabulary.string_key}"
      end

      it 'returns 204' do
        expect(response.status).to be 204
      end

      it 'removes vocabulary from database' do
        expect(Vocabulary.find_by(string_key: vocabulary.string_key)).to be nil
      end
    end
```

Commit:
Solr escaping values in fq and q queries

Commit:
Removing alt_label storing field. We cannot sort by a multivalued field.

Commit:
Changing method names from 'add_*' to 'set_*'.

Commit:
Changing test from valid uri to known uri

Commit:
Checking that custom field parameter was present in request before changing it in the record

Commit:
Moved validations to methods called by the validation step. Updating tests.

Before some validations were not going to be recorded because the valid? methods clears all errors before it starts.

Commit:
Moving local_uri_host to config/uri_service.yml. Adding generation of uri_service config file to setup rake task.

Commit:
Checking that local uris are using the host name provided in the configuration

Commit:
Improving vocabulary deletion spec
